IfcAdvancedBrep for example file cube-advanced-brep

christophKaiser opened the pull request:
In this pull request, `IfcAdvancedBrep` will be implemented to run the example file cube-advanced-brep.ifc.
This includes the implementation and adjustment of a few other functions, too. A rough overview of the involved functions was given in [a comment](https://github.com/tumcms/Open-Infra-Platform/issues/60#issuecomment-847929166) in the corresponding issue #60.

Finally, this PR will close #60.
~~Probably, issue #58 could be closed, as well. However, this will be investigated at the right time in future. [currently, the non-linked issue is intended]~~

UPDATE 21.02.2022: This PR will close #58, as well.

christophKaiser committed:
Implementation of SolidModelConverter::convertIfcAdvancedBrep

christophKaiser committed:
Added comments to SolidModelConverter::convertIfcClosedShell

christophKaiser committed:
Adjusted doxygen comment of FaceConverter::convertIfcFaceList

christophKaiser committed:
Changed to for-each-loop in FaceConverter::convertIfcFaceList

christophKaiser commented:
While I have analysed the function `convertIfcFaceList` in commit 346e611 and 032606c, I came over a 'TODO Stefan':
https://github.com/tumcms/Open-Infra-Platform/blob/1f1fc35d5c265a7c28675fe82872cd1a7595dfac/Core/src/IfcGeometryConverter/FaceConverter.h#L750-L766
(Note: The cited code is without the new changes.)

It was added in commit 2aec12f last summer. However, I wasn't able to find more information about that.
Currently, the function does its purposed task.

@pjanck Should I leave that comment there for the moment, or can you give me more information about it?

My next task in this PR will be an analyse of the function `convertIfcFace`. This will lead in a (probably) small refactoring of this function and the implementation of its subtype `IfcFaceSurface`, which is the supertype of `IfcAdvancedFace` (which I will need at the end).

pjanck commented:
Good analysis of the task!

1. About the `TODO Stefan`: no idea what I meant there. If the function does its task, you can safely delete it. 
2. You can change the `for` loop to `for( const auto& face : faces )` - perhaps that was the only purpose of the TODO?
3. `convertIfcFace` - it seems that the function returns `bool` - this can be changed to be more in parallel with other functions. I trust you here to judge by yourself what to do. Perhaps it is needed, that a `bool` is returned. In case of doubt, ask.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -748,10 +747,8 @@ namespace OpenInfraPlatform {
 
 					// Loop through all faces
-					for (auto it = faces.cbegin(); it != faces.cend(); ++it) 
+					for (const auto& face : faces) 
 					{ //TODO Stefan
-						EXPRESSReference<typename IfcEntityTypesT::IfcFace> face = (*it);
-
```
I commented before looking at the code - you've already done task 2. :)

pjanck reviewed (commented):
Good start!

christophKaiser committed:
Removed bool conversionFailed in FaceConverter::convertIfcFace.
The defined boolean value does not change inside the function convertIfcFace. Probably, its original purpose was to indicate an error inside the function. In the current code of convertIfcFace, all inspected error-cases are handled by the call of 'throw opi::<exception>'.

christophKaiser committed:
Removed variable faceId in convertIfcFace, because isn't used anymore

christophKaiser committed:
Encapsulated the loading of IfcFaceBound-list into separate function

christophKaiser committed:
Modified convertIfcFaceBoundList to avoid copy of entire vector

christophKaiser committed:
Encapsulated the loop obtainment of IfcFaceBound into separate function

christophKaiser committed:
Moved call of convertIfcFaceBound into convertIfcFaceBoundList

christophKaiser committed:
Moved external for-loop counter next to its for-loop

christophKaiser committed:
Removed not needed variable.
vertex3D and v were both declared as const carve::geom::vector<3>. By hovering with the mouse, both variables were identified as const crave::geom3d::Vector &.
Thus, I can't see any 'global transformation' like promised in the comment.

christophKaiser committed:
Fixed semi-bug of double triangulation of triangular ifcFaceBound.
See following github comment for more details on this.

christophKaiser committed:
Removed unused variable mergedIndices in convertIfcFace.
Of cause, some information was saved in the variable (line 833), but it was never used / called in the entire function.

christophKaiser commented:
[Refactoring of convertIfcFace is still in process.]

Like promised in commit 3ae4b05 “Fixed semi-bug of double triangulation of triangular ifcFaceBound”, this comment gives more details about the fixed issue – especially screenshots of an example.

Before this commit, faces with an exactly triangular bound loop (`ifcFaceBound`) were added twice to the variable `polygon` (of type `PolyhedronData`). First, the triangle face was added in the if-block `if (loopVertices3D.size() == 3)` (in `convertIfcFace`). The second adding occurred inside of the function `triangulateFace` which is called in `convertIfcFace`. This became implicitly visible by the variable `polygon.faceCount`, respectively `polygon.faceIndices`.
Moreover, the first adding has not gave respect to holes inside the outer triangle bound, respectively they were covered by the first added triangle face as shown in the following screenshot:
![screenshot fix double triangles 1 before](https://user-images.githubusercontent.com/64251870/122744124-fe12e080-d287-11eb-8c08-3a70b1419b23.png)

After this fix, the code decides between the case of only one triangular bound loop without an inner bound (= hole), and the case of an arbitrary number of vertices and / or holes inside the face. Now, the hole in my example is visible:
![screenshot fix double triangles 2 after](https://user-images.githubusercontent.com/64251870/122744144-02d79480-d288-11eb-95a4-cf12bac8e431.png)

The relevant ifc-data of my example:
```
#38= IFCFACE((#202,#203));

#202= IFCFACEOUTERBOUND(#366,.T.);
#203= IFCFACEBOUND(#367,.T.);

#366= IFCPOLYLOOP((#527,#526,#606));
#367= IFCPOLYLOOP((#528,#608,#529));

#526= IFCCARTESIANPOINT((0.,0.,0.));
#527= IFCCARTESIANPOINT((1000.,0.,0.));
#606= IFCCARTESIANPOINT((0.,1000.,0.));

#528= IFCCARTESIANPOINT((250.,250.,0.));
#529= IFCCARTESIANPOINT((500.,250.,0.));
#608= IFCCARTESIANPOINT((250.,500.,0.));
```

christophKaiser committed:
Removed unused external for-loop counter

christophKaiser committed:
Fixed parameter to passed by reference

christophKaiser committed:
Updated sub-head-line comment of including functions

christophKaiser committed:
Encapsulated process of addTriangleToPolygedronData into own function

christophKaiser committed:
Restructured convertIfcFace to decide between simple and general case:
simple case is only on triangle without holes,
general case is arbitrary number of vertices and holes

jschlenger reviewed (commented):
Good start. Excellent description of what needs to be supported in order to be able to display the example file.

christophKaiser committed:
Encapsulated process addArbitraryFaceToPolyhedronData into own function

christophKaiser committed:
Restructure if-else in convertIfcFace because of encapsulated function

christophKaiser commented:
The refactoring of the function `convertIfcFace` is done - at least for the moment. (Of course, the doxygen-comments will be updated soon.)
Maybe, its sub-function `addArbitraryFaceToPolyhedronData` could still need some effort.

Next, I will implement the functions for `IfcFaceSurface` and `IfcAdvancedFace` (both are sub-types of `IfcFace`). During or after this, I can probably decide whether [parts of] `addArbitraryFaceToPolyhedronData` could be reused there.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -726,7 +726,8 @@ namespace OpenInfraPlatform {
-				// convertIfcFaceList, convertIfcFace, convert3DPointsTo2D, triangulateFace, convertIfcCartesianPointVectorVector
+				// convertIfcFaceList, convertIfcFace, convertIfcFaceBoundList, covnertIfcFaceBound, addTriangleToPolyhedronData, 
+				// addArbitraryFaceToPolyhedronData, convert3DPointsTo2D, triangulateFace, convertIfcCartesianPointVectorVector
 				//------------------------------------------------------------------------------------------------------------
```
This can safely be deleted, it adds no value.

christophKaiser replied:
Done in commit 93d2916.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -748,15 +749,13 @@ namespace OpenInfraPlatform {
+					{
 						if (face.expired()) {
 							throw oip::ReferenceExpiredException(face);
 						}
```
This should be checked as the first thing in the `convertIfcFace` function, no need to do it here.

christophKaiser replied:
Done in commit b6148c3.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -768,132 +767,165 @@ namespace OpenInfraPlatform {
+				{
+					// As carve expects outer boundary of face to be at first index, outer boundary has to be moved to index 0 and inner boundary has index >= 1
+					for (auto it = ifcFaceBounds.begin(); it != ifcFaceBounds.end(); it++)
+					{
+						if (it->isOfType<typename IfcEntityTypesT::IfcFaceOuterBound>())
+						{
+							// swap element 'it' with first element
+							std::rotate(ifcFaceBounds.begin(), it, it + 1);
 
-					// Loop through all boundary definitions
-					int boundID = -1;
+							// according to ifc-documentation, only one boundary (= loop) can be outer boundary:
+							// "One loop is optionally distinguished as the outer loop of the face."
+							// (https://standards.buildingsmart.org/IFC/DEV/IFC4_3/RC3/HTML/link/ifcface.htm)
+							// thus, break the for-loop after THE outer boundary was found
+							break;
+						}
+					}
```
I propose to make this a distinct function.

Takes in `std::vector<EXPRESSReference<... IfcFaceBound>>& faceBound` and puts the outer bound at index 0.

This way, the input to `convertIfcFaceBoundList`  can be made const reference (`const ...&`).

christophKaiser replied:
Done in commit b4c24cb.
However, the input of the new distinct function `swapOuterBoundaryToFront` can't be a reference to a constant vector because the function `std::rotate(...)` performs changes on this vector.

An alternative approach to avoid the operation on the vector (and prevent the copy with the swap of elements) could be to only indicate the index of the outer boundary. In the function `convertIfcFaceBoundList`, the outer boundary would be added as first element to the target `faceBoundLoops`, and then the for-loop would add all other loops (with skipping the outer boundary).
At least in terms of comprehensible code, probably this would be more complex.

pjanck replied:
not const is perfect, simple code is preferred :)

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -768,132 +767,165 @@ namespace OpenInfraPlatform {
+				void convertIfcFaceBoundList(
+					std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> ifcFaceBounds,
+					const carve::math::Matrix& pos,
+					std::vector<std::vector<carve::geom::vector<3>>>& faceBoundLoops) const noexcept(true)
```
Perhaps the `faceBoundLoops` can be made a return parameter?

christophKaiser replied:
Done in commit d9303cc.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -768,132 +767,165 @@ namespace OpenInfraPlatform {
+				void convertIfcFaceBound(
+					const EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>& bound,
+					const carve::math::Matrix& pos,
+					std::vector<carve::geom::vector<3>>& loopVertices3D) const noexcept(false)
```
Similarly to above, `loopVertices3D` can be made return parameter?

christophKaiser replied:
Done in commit ff1b65e.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -768,132 +767,165 @@ namespace OpenInfraPlatform {
+					const EXPRESSReference<typename IfcEntityTypesT::IfcLoop>& loop = bound->Bound;
+					bool polyOrientation = bound->Orientation;
+
+					if (!loop) {
```
test, if reference expired

christophKaiser replied:
Done in commit 785881d.
I added the expired-test of the `EXPRESSReference<... IfcFaceBound> bound` because it wasn't done before.
As well, I adjusted and kept the expired-test of the `EXPRESSReference<... IfcLoop> loop`, which is done immediately after it's definition. At this point, I'm not sure about it - I would expect this test at the beginning of the function `CurveConverter::convertIfcLoop`, however I can't spot it there.

pjanck replied:
if the test is not at the beginning of the function, please put it there

christophKaiser replied:
Done in commit 5d90a20.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -768,132 +767,165 @@ namespace OpenInfraPlatform {
 
-					for (const auto& bound : modBounds) 
+					if (loopVertices3D.size() < 3)
 					{
-						boundID++;
+						throw oip::InconsistentGeometryException(loop, " Number of vertices < 3");
+					}
```
You can test this before the for-loop applying the position.

christophKaiser replied:
Done in commit 72e884d.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -768,132 +767,165 @@ namespace OpenInfraPlatform {
-						//	IfcLoop <- IfcEdgeLoop, IfcPolyLoop, IfcVertexLoop
-						const EXPRESSReference<typename IfcEntityTypesT::IfcLoop>& loop = bound->Bound;
-						bool polyOrientation = bound->Orientation;
+					return;
+				}
```
No need for a `return` statement at the end of a `void` function (it happens nevertheless).

christophKaiser replied:
Indirectly done in commit ff1b65e, becouse the function type was changed from `void` to a vector.
Done for other return statements in commit 3f4c077.

pjanck reviewed (commented):
Some suggestions for improvement. Otherwise, seems fine - good progress!

jschlenger reviewed (commented):
Not able to spot any additional mistakes

christophKaiser committed:
Removed subtitle comment

christophKaiser committed:
Removed valid-check of EXPRESSReference entity (weak pointer).
This check is the first task in the function convertIfcFace which is called at this place.

christophKaiser committed:
Moved 'swap of outerBoundary' into a distinct function

christophKaiser committed:
Changed return type of convertIfcFaceBoundList

christophKaiser committed:
Changed return type of convertIfcFaceBound

christophKaiser committed:
Added test of reference expired to convertIfcFaceBound.
Adjusted expired test of IfcLoop.

christophKaiser committed:
Moved integrity test about number of loop points upwards

christophKaiser committed:
Removed not needed return statements

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -801,59 +790,73 @@ namespace OpenInfraPlatform {
+					return faceBoundLoops;
+				}
+
+				std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> swapOuterBoundaryToFront(
+					std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> ifcFaceBounds) const throw(...)
```
if you're doing swaps in place, no need to return, use reference `&`
```suggestion
				void swapOuterBoundaryToFront(
					std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>>& ifcFaceBounds) const throw(...)

```

christophKaiser replied:
Done in commit cc2b38a.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -801,59 +790,73 @@ namespace OpenInfraPlatform {
+
+					std::vector<std::vector<carve::geom::vector<3>>> faceBoundLoops;
+					faceBoundLoops.resize(faceBounds.size());
+					for (size_t i = 0; i < faceBounds.size(); i++)
+					{
+						// get vertices of one bound loop into faceBoundLoops[i]
+						faceBoundLoops[i] = convertIfcFaceBound(faceBounds[i], pos);
+					}
```
you could use `std::transform` here (but not necessary to change; you can learn, however)

christophKaiser replied:
Done in commit 7096ed8.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -801,59 +790,73 @@ namespace OpenInfraPlatform {
-						throw oip::InconsistentModellingException(bound, " No valid loop");
+					if (loop.expired()) {
+						throw oip::ReferenceExpiredException(loop);
 					}
```
check in `convertIfcLoop`

christophKaiser replied:
Done in commit 5d90a20

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -801,59 +790,73 @@ namespace OpenInfraPlatform {
+					std::vector<carve::geom::vector<3>> loopVertices3D;
+
 					// Collect all vertices of the current loop
 					curveConverter->convertIfcLoop(loop, loopVertices3D);
```
perhaps change the return value of the function?

christophKaiser replied:
Done in commit 8cc1112.
The function `convertIfcEdgeLoop` will receive some futher attention within this pull request.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -801,59 +790,73 @@ namespace OpenInfraPlatform {
+
 					//	IfcLoop   has subtypes   IfcEdgeLoop, IfcPolyLoop, IfcVertexLoop
 					const EXPRESSReference<typename IfcEntityTypesT::IfcLoop>& loop = bound->Bound;
 					bool polyOrientation = bound->Orientation;
```
do we react to this boolean?

christophKaiser replied:
A few lines later, it's used to flip the orientation / sense of the boundary.
https://github.com/tumcms/Open-Infra-Platform/blob/2715f2c87179552ecdd25ba656edf031af0fa9cb/Core/src/IfcGeometryConverter/FaceConverter.h#L849-L853
But the variable is only used once, thus commit 3aab188 would replace the variable with its actual information source.

pjanck reviewed (changes requested):
small suggestions

christophKaiser committed:
Merge branch 'development' into 60AdvancedBrep

christophKaiser committed:
Moved ifcLoop.expired() from convertIfcFaceBound into convertIfcLoop.
Adjusted variable name according to the lowerCamelCase convention.

christophKaiser committed:
Changed return type of convertIfcLoop and its sub-functions

christophKaiser committed:
replaced variable bool polyOrientation by its information source

christophKaiser committed:
Changed swapOuterBoundaryToFront to work with reference.
The input std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> ifcFaceBounds of the function convertIfcFaceBoundList is changed to non-const non-reference, because at some point the original const source has to be converted to non-const (in order to perform the swap).

christophKaiser committed:
Replaced for-loop with std::transform

christophKaiser committed:
Added doxygen comments to convertIfcFaceBound (and related functions)

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -790,43 +790,60 @@ namespace OpenInfraPlatform {
+				 */
 				std::vector<std::vector<carve::geom::vector<3>>> convertIfcFaceBoundList(
-					const std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>>& ifcFaceBounds,
+					std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> ifcFaceBounds,
```
I'd leave the reference `&` - this way the vector is not unnecessarily copied...

christophKaiser replied:
With the reference `&`, the variable `ifcFaceBounds` becoms a const vector, because the source `face->Bounds` in `convertIfcFace` (line 775) is a const variable. (`const EXPRESSReference<typename IfcEntityTypesT::IfcFace>& face`, line 763)
At least, the function `swapOuterBoundaryToFront` (called in line 805) needs a non-const vector in order to perform the swap of elements. Probably, we need the copy at some point to come from const to non-const? (see message of commit cc2b38a)
By just adding the `&` in line 802 (where you commented), the compiler complains "cannot convert argument 1 from 'const (...)' to 'std::vector<(...)> &' "

pjanck replied:
Right, then it needs copying, I see. I withdraw my comment.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -790,43 +790,60 @@ namespace OpenInfraPlatform {
-						// get vertices of one bound loop into faceBoundLoops[i]
-						faceBoundLoops[i] = convertIfcFaceBound(faceBounds[i], pos);
-					}
+					faceBoundLoops.resize(ifcFaceBounds.size());
```
This can done in a single line, to reserve the amount of space in a vector, I think it is this syntax:

`std::vector<std::vector<carve::geom::vector<3>>> faceBoundLoops(ifcFaceBounds.size());`

christophKaiser replied:
Thanks for this hint! I had forgot about that.
Done in commit 3f00fd6.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -790,43 +790,60 @@ namespace OpenInfraPlatform {
+				 * Only one boundary loop can be an \c IfcFaceOuterBound entity. But in general, it's optional. 
+				 * \param[in,out]	ifcFaceBounds	List of \c IfcFaceBound entities in which the \c IfcFaceOuterBound entity has to be swaped to the front.
+				 */
+				void swapOuterBoundaryToFront(
+					std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>>& ifcFaceBounds
+				) const throw(...)
```
suggestion:

throw if there are more than 1 outer boundaries - you can use `std::count_if`

christophKaiser replied:
`throw` is introduced with commit a9990e7.
I'm not sure about the throw message: Usually, the custom `throw`s expect a single `EXPRESSReference` entity (if I understood it right). However, at this point the function has a bunch (vector) of entities which causes the throw. Just to give the user a hint about from where the error comes, I put the id of at least one entity into the string-message overload.
When trace back to a single source entity, the `IfcFace` in the function `convertIfcFace` would be the appropriate one - but currently, the function doesn't have access to it. To add it as a parameter just for the throw feels like not the best choice.

pjanck replied:
Good analysis. No need for a special parameter for a `throw`. Your solution seems fine.

pjanck reviewed (changes requested):
Some suggestions

christophKaiser commented:
Currently, the function `convertIfcFaceBound` and its superfunction `convertIfcFaceBoundList` are implemented in the class `FaceConverter` and are called only by the function `convertIfcFace`.
However, the entity `IfcFaceBound` is a subtype of `IfcTopologicalRepresentationItem` (according to the [documentation](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_1/FINAL/HTML/link/ifcfacebound.htm)). Thus, should they be implemented in the class `ReprestationConverter` (shift the existing code into that place) and be called by the function `convertIfcTopologicalRepresentationItem` as well?

pjanck commented:
- no need to move the code, it is fine where it is (the functions are converting faces, right?)
- call appropriately from the `convertIfcTopologicalRepresentationItem` function - representationConverter has a faceConverter as a member (see e.g. the `convertIfcGeometricRepresentationItem`)

christophKaiser committed:
Fixed doxygen comment of convertIfcTopologicalRepresentationItem

christophKaiser committed:
Added IfcFace to convertIfcTopologicalRepresentationItem

christophKaiser committed:
Added IfcFaceBound to convertIfcTopologicalRepresentationItem

christophKaiser committed:
Added IfcLoop to convertIfcTopologicalRepresentationItem

christophKaiser committed:
Fix: Compressed vector declaration

christophKaiser commented:
> 
> 
>     * call appropriately from the `convertIfcTopologicalRepresentationItem` function - representationConverter has a faceConverter as a member (see e.g. the `convertIfcGeometricRepresentationItem`)

Done in commit 5c2770c (`IfcFaceBound`).
I have observed the same issue with `IfcFace` and `IfcLoop`. The calls are added similar in commits 54e028b and 01a3edc.

pjanck commented on Core/src/IfcGeometryConverter/RepresentationConverter.h:
```diff
@@ -510,6 +510,59 @@ namespace OpenInfraPlatform {
+						polyline_data->beginPolyline();
+
+						// add all vertices to polyline
+						for (auto v : vertices) {
```
```suggestion
						for (const auto& v : vertices) {
```

I'm not sure this would work, please try.

christophKaiser replied:
If work == compiles: Done in commit 363c1cf.
However, I wasn't able to run this code block with an ifc file. Probably, I will create one for an unit test of `IfcTopologicalRepresentationItem`'s a bit later (especially to test the throw statements inside the called functions and the issue of an `IfcFace` with hole which I observed [above](https://github.com/tumcms/Open-Infra-Platform/pull/462#issuecomment-864908637)).

pjanck commented on Core/src/IfcGeometryConverter/RepresentationConverter.h:
```diff
@@ -510,6 +510,59 @@ namespace OpenInfraPlatform {
+						polyline_data->beginPolyline();
+
+						// add all vertices to polyline
+						for (auto v : vertices) {
```
similarly

christophKaiser replied:
Done in commit 363c1cf.

pjanck reviewed (commented):
Good improvements!

pjanck commented:
> I have observed the same issue with IfcFace and IfcLoop. 

Very good! If you wish, you can look into `IfcVertex` as well. Should be easy to put together and some functionality should be already present in the converters.

christophKaiser committed:
Added throw in swapOuterBoundaryToFront.
The function throws, if there are multiple IfcFaceOuterBoundary entities.

christophKaiser committed:
Specified auto-type declaration more specifically with const &.

christophKaiser commented:
> Very good! If you wish, you can look into `IfcVertex` as well. Should be easy to put together and some functionality should be already present in the converters.

`IfcVertex` with `IfcVertexPoint` are already on my [agenda](https://github.com/tumcms/Open-Infra-Platform/issues/60#issuecomment-847929166) - they are used by the example file cube-advanced-brep. I will come to them after some effort into `IfcLoop` (with subtypes).

christophKaiser committed:
Added conditional construct for IfcFaceSurface into convertIfcFace

christophKaiser committed:
Added body and call of convertIfcFaceSurface

christophKaiser committed:
Added body and call of convertIfcAdvancedFace

christophKaiser committed:
Added check of IfcFaceBound-types in convertIfcAdvancedFace

christophKaiser committed:
Added check of IfcSurface-types in convertIfcAdvancedFace

christophKaiser committed:
Added function call to calculate IfcPlane in convertIfcAdvancedFace

christophKaiser committed:
Added call to IfcBSplineSurface in convertIfcAdvancedFace. Note:
Surface is not append in return variable.

christophKaiser committed:
Added mergePolyhedronsIntoOnePolyhedron to append surface to result

christophKaiser commented:
The current pushes implement the functions to convert an `IfcAdvancedFace`. Surprisingly, the visual result does not look wrong:

![210924 MidtermResult cube](https://user-images.githubusercontent.com/64251870/134642464-33cf1641-1187-4609-af08-ac3e86bfdea3.png)
![210924 MidtermResult basin](https://user-images.githubusercontent.com/64251870/134642511-5f14fc37-6f93-4a98-81b5-67e98ee27114.png)

I'm surprised because there is a bunch of functions and ifc-entities on my agenda, where I want to take a detailed look into - for example `convertIfcLoop`. Maybe, they actually do the right things, or it's just luck to get right looking results. Furthermore, in those subfunctions a few ToDo's have been there before the start of this pull request.

Besides that, in the current work are a few things, which might need some rework and additions, like:
- take `sameSense` into account by the `addFace`-construction
- the B-Spline surface should be trimmed by the loop boundary: I don't have an idea how to do this task, but I already have a few papers which talk about it.
- discussion about the return type of `convertIfcSurface`: Each surface (e.g., each side of a cube) is calculated into one `Polyhedron`. According to my general understanding of a polyhedron, the entire cube should be stored in one `Polyhedron`-object. Currently, the new function `mergePolyhedronsIntoOnePolyhedron` copies the data into one object.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -774,20 +774,124 @@ namespace OpenInfraPlatform {
+					std::vector<std::vector<carve::geom::vector<3>>>& faceBoundLoops,
+					const carve::math::Matrix& pos,
+					std::shared_ptr<carve::input::PolyhedronData>& polygon, //Carve polygon of the converted face
+					std::map<std::string, uint32_t> polygonIndices // Contains polygon indices of vertices (x,y,z converted to string)
```
```suggestion
					std::map<std::string, uint32_t>& polygonIndices // Contains polygon indices of vertices (x,y,z converted to string)
```

Don't copy, pass by reference. (apply to other functions as well)

christophKaiser replied:
Done in commit 2034c1c.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -774,20 +774,124 @@ namespace OpenInfraPlatform {
+
+					// get reference to IfcFaceBound's (attribute 1)
+					const std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>>& ifcFaceBounds = advancedFace->Bounds;
+					// if one face bound is not of the allowed type: throw; else: continue and use already converted faceBoundLoops
+					if (!std::all_of(ifcFaceBounds.begin(), ifcFaceBounds.end(), [](const auto& ifcFaceBound) {return
+						ifcFaceBound->Bound.isOfType<typename IfcEntityTypesT::IfcEdgeLoop>() || ifcFaceBound->Bound.isOfType<typename IfcEntityTypesT::IfcVertexLoop>(); }))
```
Genial!

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -774,20 +774,124 @@ namespace OpenInfraPlatform {
+					// ToDo: take sameSense into account by the addFace-construction
+					const bool sameSense = advancedFace->SameSense;
+
+					// get attribute 2
+					const EXPRESSReference<typename IfcEntityTypesT::IfcSurface>& surface = advancedFace->FaceSurface;
```
Is there no `convertIfcSurface` available? If not, please introduce one.

christophKaiser replied:
This section is moved into the new function computeIfcFaceSurface (commit b9d8d5d).
Inside there, the surface is needed as EXPRESSReference, because its surface-type (IfcPlane or everything else) is part of the decision about the computation approach. The big if-else construct was simplified in commit 6c3e76d.
Due to this simplification, the existing function `convertIfcSurface` is called.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -774,20 +774,124 @@ namespace OpenInfraPlatform {
+						// ASSUMPTION: the B-Spline-Surface boundary is already coincident with the loop boundary
+						// ToDo: in general case, the B-Spline-Surface has to be trimmed by the loop boundary (or at least checked to be coincident)
+
+						// append surface-faces to target polygon
+						mergePolyhedronsIntoOnePolyhedron(inputDataFaceSurface, polygon);
```
Please make this a member function of `ItemData`.

christophKaiser replied:
Done in commit 237f2b3.

pjanck reviewed (commented):
I'm impressed with the screenshots!

christophKaiser committed:
Fixed function parameter to be reference

christophKaiser committed:
Adjusted variable name of the FaceSurface-entity to be more meaningful

christophKaiser committed:
Adjusted parameter names in mergePolyhedronsIntoOnePolyhedron

christophKaiser committed:
Moved mergePolyhedronsIntoOnePolyhedron into class ItemData

christophKaiser committed:
Fixed initialization of temporal ItemData variable

christophKaiser committed:
Introduced function computeFaceSurface, moved content ...
from convertAdvancedFace into computeFaceSurface.
The new function computeFaceSurface will be called from two functions (convertIfcFaceSurface and convertAdvancedFace) - call is coming soon.

christophKaiser committed:
Changed check of surface-type restriction

christophKaiser committed:
Added call of computeIfcFaceSurface in convertIfcFaceSurface

christophKaiser committed:
Changed variable name polygon to polyhedron in convertFace and following
Actually, a polyhedron (3D object) is constructed with many polygons (2D face with arbitrary number of corner points). The variable name "polygon" confused me when I was thinking about how to glue several polygons to one polyhedron by topology.

christophKaiser committed:
Removed the parameter faceBoundLoops.
Now, it's loaded when it will be needed (2 times).

christophKaiser committed:
mergePolyhedronsIntoOnePolyhedron takes existing points into account.
Previous, several polyhedrons were just put together into one polyhedron but without a topological merge at the edges (just geometrical coincident edges).
Now, the functions searches whether the points already exit inside the target polyhedron; if they exist, the polyhedrons will be merged topological (by reuse of the exiting point index).

christophKaiser committed:
Added checkPointIsPartOfSurface in computeIfcFaceSurface.
The multiple nested for-loop is split: One part is replaced with std::all_of, the other part is separated into a new sub-function.

christophKaiser committed:
Added checkBoundaryIsPartOfSurface in computeIfcFaceSurface.

christophKaiser committed:
Added doxygen-comments to convertIfcFace and sub-functions

christophKaiser commented:
Beside small adjustments (commits 2034c1c, bc916f2, b59dcb3, 100c1e1, ef67c56), the structure of functions which are called from `convertIfcFace` was reordered:
A new function `computeIfcFaceSurface` was introduced in commit b9d8d5d. This function is separated, because it is called by `convertIfcFaceSurface` and `convertIfcAdvancedFace` (see calls in commits b9d8d5d and 25d2800) - in both cases it has to do the same computation.
In commit 86f6db6, the parameter `faceBoundLoops` was removed and replaced by the appropriate function call at the location when it is needed.
Previously, the function `mergePolyhedronsIntoOnePolyhedron` has put together several polyhedrons into one polyhedron, but without a topological merge. The result was like a few LEGO bricks in one bowl without sticked together. Now with commit 88e11fe, the existing mesh points are taken into account. If a point (respectively its coordinates) already exists in the target mesh, the existing point will be reused - in this way, several meshes are connected topological. In the example above, the several LEGO bricks in the bowl are sticked together.
Doxygen-comments were added in commit 97efdce.

Commits 34b0b46 and 1cb442c added two functions check whether the points of the boundary loop exist in the surface mesh. If not all boundary loop points are in the surface, a warning is displayed in the console output. Its purpose is to indicate, whether the calculated geometry could be wrong (but don't has to be) - especially in the case when the surface should be trimmed by the boundary (but it isn't, due to a missing functionality).
The example file basin-advanced-brep.ifc works fine with that. However, this warning is raised in the example cube-advanced-brep.ifc, where it shouldn't. It's caused by a default precision of 1.49e-08 (from GeomSettings->getPrecision, fix value in carve), which is to tight. If the precision of 1.0e-4 from this specifc ifc-file is used (defined in the entity `IfcGeometricRepresentationContext`), this problem will not occur - an appropriate ToDo is already in the code:
https://github.com/tumcms/Open-Infra-Platform/blob/025cf1713a7d8a39d24ac58822d7b8c7e7a1aa03/Core/src/IfcGeometryConverter/GeometrySettings.cpp#L102

Beside this aesthetical issue, this pull request has reached its intended functionality (as communicated in dev-meetings). As indicated in the previous section, the trimming of the face surface by the boundary loop is not implemented. This problem will be explained in more detail in a new issue.
Furthermore, the attribute sameSense (see in function computeIfcFaceSurface) is not considered, an appropriate ToDo is left in the code. This attribute affects the normal direction of the faces. At least with the current shading settings in OIP, the normal direction has no visible impact.

Before merge, three ToDo's left on my side:
- [x] Merge `tumcms/development` (with update to VS2019) into this branch, and make sure it works with VS2019 [at last done in commit a83a47b]
- [x] Add unit tests [done in commits 41e84a4, 14e2007 and bbceed0]
- [x] Update the list in SupportedIFCrepresentations.md [done in commit df133e2]

UPDATE 21.02.2022: Updated check boxes of ToDo's.

christophKaiser committed:
Merge branch 'development' into '60AdvancedBrep'

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -726,85 +726,89 @@ namespace OpenInfraPlatform {
+				\param[in]		face				\c IfcFace entity to be converted.
+				\param[in]		pos					The relative location of the origin of the representation's coordinate system within the geometric context.
+				\param[in,out]	polyhedron			Carve polyhedron of the converted face.
+				\param[in,out]	polyhedronIndices	Contains polyhedron indices of vertices (x,y,z converted to string).
```
Suggestion: I think Doxy allows for tags like `<see ... >` where you can link to a specific other function in the documentation. Perhaps worth a try here to link to the function that converts the indices?

christophKaiser replied:
At the moment, there is no explicit function which converts the `polyhedronIndices`. However, thanks for your comment - it showed me the need of refactoring at this point. For another programmer, it isn't easy to see from where the variable `polyhedronIndices` comes from, how it has to be initialised, does it need some kine of default pre-set, and so on.

**A few words about how it works right now**
The variable `std::map<std::string, uint32_t> polyhedronIndices` is declared at one place - in this case inside the function `convertIfcFaceList`. From there it's passed by reference to all functions which work with it.

'Work' refers to the adding a vertex to `polyhedron`. In that case, a `stringstream` with the vertex coordinates is formed and searched inside of `polyhedronIndices`. If the `stringstream` exists, the stored vertex index is used. Otherwise, the vertex is added to `polyhedron` and the `stringstream` is added to `polyhedronIndices`. This task is done by the following sample code:

```cpp
// set string id and search for existing vertex in polyhedron
std::stringstream vertexString;
vertexString << vertex3D.x << " " << vertex3D.y << " " << vertex3D.z;
auto itFound = polyhedronIndices.find(vertexString.str());

uint32_t index = 0;
if (itFound != polyhedronIndices.end()) {
	index = itFound->second;
}
else {
	index = polyhedron->addVertex(vertex3D);
	polyhedronIndices[vertexString.str()] = index;
}

// use index (depends on context)
```

The variable polyhedronIndices is passed by reference through the multiple function calls because its content is filled incrementally and reused in the next calls. In this way, it's responsible for the topological merge of coincident vertices.

The above sample code can be found at multiple places throughout the project - sometimes with little adaptions:
- Subfunctions of `FaceConverterT::convertIfcFaceList`:
    - `FaceConverterT::addTriangleToPolyhedronData`
    - `FaceConverterT::triangulateFace`
    - `ItemData::mergePolyhedronsIntoOnePolyhedron`
- Local use in `FaceConverterT::convertIfcBSplineSurface`
- Local use in `ProfilConverterT::SearchExistingVertex`
- Local use in `SolidModelConverterT::convertIfcSpecificSolidModel` (function is commented out)

**Thoughts about how to refactor**
Because the same task is done at multiple places in the project, it should be reduced into one central function. Within this change, the variable `polyhedronIndices` should be somehow bound to the corresponding polyhedron, hence another programmer don't has to care that much about this variable.

As one conclusion, the information in `polyhedronIndices` is just another representation of the content from the member variable `std::vector<carve::geom3d::Vector> points` of the class `carve::input::PolyhedronData`, respectively `carve::input::PolylineSetData` (`points` is inherited from the class `carve::input::VertexData` where it's declared). Probably, the type `std::map<std::string, uint32_t>` was choosen to have access the `find` function of `std::map`.

To be accurate, `polyhedronIndices` should be a member variable in `VertexData` and the above samle code becomes a memberfunction inside there. However, this is outside of our actual namespace.

Inside the namespace OIP, the class `ItemData` would be the next alternative for the member variable and its depending functions - the function `mergePolyhedronsIntoOnePolyhedron` which works within this topic of topological merging is already there. The first idea was to keep and use the existing instances of `PolyhedronData` and `PolylineSetData`. One disadvantage of this was, that we would need one `polyhedronIndices`-variable for each instance of `PolyhedronData` and `PolylineSetData`, thus four vectors of them (for `closed_polyhedrons`, `open_polyhedrons`, `open_or_closed_polyhedrons` and `polylines`). Futhermore, assuming on function is in execution on one instance of a polyhedron, it will be quite difficult to determine the corresponding `polyhedronIndices`-variable.

A second thought on resolving the determination issue would be the introduction of a new polyhedron instance / variable with its corresponding `polyhedronIndices`-variable. The job of this new instance would be to collect the geometry of one polyhedron; once the polyhedron is complet, its pointer would be handed over into the existing vector `closed_polyhedrons`. The sample code from above would be in the same class, where it can be explicitly called when this functionality should be used; otherwise this variables and functions about topological merging don't have to be called (like in the majority of converter functions which don't use this search methodology).
For the implementation of the new variables and functions could be two options: At first, an extension of the existing class `ItemData` would be an option. The disadvantage on this would be that the new variables and functions would be always visible in an `ItemData`-object.
The second option would be a new class which is used similar to the `ItemData`-class during the collection of the polyhedron. If the collection is finished, the polyhedron would be handed over into the target `ItemData`-object. About the parameter of the converter functions, one issue in this option could be the compatibility between `ItemData`-objects and objects of the new class.

christophKaiser replied:
With reference to todays meeting, the described problem will be subject of a new github-issue.
However, I want to update the doxygen comment to make more clear where the parameter `polyhedronIndices` comes from and how to handle it.

christophKaiser replied:
Done.
- New github-issue with the described problem is created.
- Doxygen comments are updated by commit ac64eb1; the result is shown in the comment [below](https://github.com/tumcms/Open-Infra-Platform/pull/462#issuecomment-1046816351).

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -726,85 +726,89 @@ namespace OpenInfraPlatform {
+				\param[in]		faceSurface			\c IfcFaceSurface entity to be converted.
+				\param[in]		pos					The relative location of the origin of the representation's coordinate system within the geometric context.
+				\param[in,out]	polyhedron			Carve polyhedron of the converted face.
+				\param[in,out]	polyhedronIndices	Contains polyhedron indices of vertices (x,y,z converted to string).
```
simile.

christophKaiser replied:
Done.
Doxygen comments are updated by commit ac64eb1; the result is shown in the comment [below](https://github.com/tumcms/Open-Infra-Platform/pull/462#issuecomment-1046816351).

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -813,20 +817,25 @@ namespace OpenInfraPlatform {
 					if (faceSurface.isOfType<typename IfcEntityTypesT::IfcAdvancedFace>())
 					{
-						convertIfcAdvancedFace(faceSurface.as<typename IfcEntityTypesT::IfcAdvancedFace>(), faceBoundLoops, pos, polygon, polygonIndices);
+						convertIfcAdvancedFace(faceSurface.as<typename IfcEntityTypesT::IfcAdvancedFace>(), pos, polyhedron, polyhedronIndices);
```
Doesn't the compiler complain about the missing `template` keywords at function calls?

christophKaiser replied:
On my system, the compiler has not serious issue with this. I'm albe to run the UI-Project and load the example file with an IfcAdvancedFace, as well the Unit Tests run without problems.
However, I still have VS2017 installed on my system - I hope this doesn't affect the behavior of the compiler.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -835,65 +844,131 @@ namespace OpenInfraPlatform {
+				\param[in,out]	polyhedron			Carve polyhedron of the converted face.
+				\param[in,out]	polyhedronIndices	Contains polyhedron indices of vertices (x,y,z converted to string).
+				*/
+				void computeIfcFaceSurface(
```
We have usually used `convert...` for function names that convert Ifc content to carve. For the sake of consistency, please change.

christophKaiser replied:
There is already a function `convertIfcFaceSurface` within this pull request, about in line 807.
The idea behind: `convertIfcFaceSurface` will be called from `convertIfcFace` or from somewhere else. Depending on the ifc-entity, that function calls `convertIfcAdvancedFace` or hands over straight to this function `computeIfcFaceSurface`. As well, `convertIfcAdvancedFace` will hand over to `computeIfcFaceSurface`. `computeIfcFaceSurface` performes the actual calculation for both cases, IfcFaceSurface and IfcAdvancedFace. In the visualisation of doxygen, it looks the following:
![computeFaceSurface_callStructure](https://user-images.githubusercontent.com/64251870/146552738-18e22b56-7636-4eac-afae-10a94f14bdd2.png)
If we would apply the distinction of public and private functions in the geometry converter classes, `convertIfcFaceSurface` and `convertIfcAdvancedFace` would be public, and `computeIfcFaceSurface` would be private.

pjanck replied:
Oh, I see. Please, do use the public/private distinction. You can put the function at the end of the file as a private function - this is already used in some places.

christophKaiser replied:
Done in commit 0e39d35. I moved it into the protected section, in case another function will inherit from `FaceConverterT`.
Further functions moved to protected with commits 332747a and d6dbb6d.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -835,65 +844,131 @@ namespace OpenInfraPlatform {
+						// ToDo: in general case, the Surface has to be trimmed by the loop boundary (or at least checked to be coincident)
+						if (!checkBoundaryIsPartOfSurface(inputDataFaceSurface, faceBoundLoops))
+						{
+							BLUE_LOG(warning) << "Surface of IfcFaceSurface (or IfcAdvancedSurface) does not meet the boundary, or the default precision is to tight. Geometry could be wrong!";
```
```suggestion
							BLUE_LOG(warning) << "Surface of IfcFaceSurface (or IfcAdvancedSurface) does not meet the boundary, or the default precision is too tight. Geometry could be wrong!";
```

christophKaiser replied:
Done in commit 9cd94a5

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -835,65 +844,131 @@ namespace OpenInfraPlatform {
+					return std::all_of(faceBoundLoops.begin(), faceBoundLoops.end(), [&, this](const auto& faceBoundLoop) {
+						return std::all_of(faceBoundLoop.begin(), faceBoundLoop.end(), [&, this](const auto& pointLoop) {
+							return checkPointIsPartOfSurface(itemDataSurface, pointLoop); });
+					});
```
Awesome!!!

pjanck reviewed (changes requested):
Some suggestions and praise. Waiting for the rest of the Todos to merge.

christophKaiser committed:
Merge branch 'development' into 60AdvancedBrep

christophKaiser committed:
Fixed typo

christophKaiser committed:
Moved function computeIfcFaceSurface into protected section. ...
... It's an internal sub-function of convertIfcFaceSurface and convertIfcAdvancedFace.

christophKaiser committed:
Moved further function into protected section. ...
The functions addTriangleToPolyhedronData, addArbitraryFaceToPolyhedronData, convert3DPointsTo2D, and triangulateFace are internal sub-functions of convertIfcFace.

christophKaiser committed:
Moved function swapOuterBoundaryToFront into protected section. ...
... It's an internal sub-function of convertIfcFaceBoundList.

christophKaiser committed:
Merge branch 'development' into 60AdvancedBrep

pjanck reviewed (commented):
Good work so far!

christophKaiser committed:
Deleted a Doxygen note which legacy from previous the refactoring.
At least I have no idea which information or benefit the note should give to the user or programmer.

christophKaiser committed:
Added Doxygen description of variables polyhedron and polyhedronIndices
The description is placed in the comment of convertIfcFace.
Added 'See also' (\see) to all functions inside the FaceConverter which have the parameter polyhedron and polyhedronIndices.

jschlenger commented on Core/src/IfcGeometryConverter/RepresentationConverter.h:
```diff
@@ -513,9 +513,9 @@ namespace OpenInfraPlatform {
-				 * \param[in] mapped_item The \c IfcTopologicalRepresentationItem to be converted.
-				 * \param[in] pos The relative location of the origin of the representation's coordinate system within the geometric context.
-				 * \param[out] itemData A pointer to be filled with the relevant data.
+				 * \param[in] topo_item			The \c IfcTopologicalRepresentationItem to be converted.
+				 * \param[in] objectPlacement	The relative location of the origin of the representation's coordinate system within the geometric context.
+				 * \param[out] itemData			A pointer to be filled with the relevant data.
```
I know "pos" is maybe not the best name to understand what is meant but I'm quite sure that pos is used in a lot of converter functions 
So having multiple names for the same thing might not be the best idea 
But it's just a minor comment and I leave it up to you if you want to change it

pjanck replied:
My vote goes for: let's leave at `pos`.

christophKaiser replied:
I see the point of the more understandable name `objectPlacement`.
With a rough search, VS found 241 matches in 6 files (.Core project). In case we want to change this name, a new PR would be suitable in my opinion.
Anyway, I have made a note to not forget this in the next meeting where we could talk more about this?

pjanck replied:
We can definitely change this name, but rather in a new PR.

christophKaiser replied:
An issue as reminder and basis for a new PR is created.

jschlenger reviewed (approved):
Looks very good to me.
We should merge this pull request soon.

pjanck commented:
Please mark it as ready for review, then I'll go through the code once more.

christophKaiser commented:
Commit ac64eb1 updates the Doxygen comments, which are related to the parameters `polyhedron` and `polyhedronIndices`.

A detailed explanation about these parameters is in the description of the function `convertIfcFace`. As shown in the screenshot below, this section is separated with a bold header:
![doxygen_description](https://user-images.githubusercontent.com/64251870/154952590-a9085fb5-a09b-4b41-abac-237256b1f872.png)

Furthermore, all functions which have `polyhedron` and `polyhedronIndices` as parameter refer to this description by the Doxygen command `\see`. This generates the section 'See also', as shown in the screenshot:
![doxygen_seeAlso](https://user-images.githubusercontent.com/64251870/154952564-8006a4b9-e39b-438a-90bf-806534b53daf.png)

christophKaiser committed:
Added unit test for IfcFace
(The example file face.ifc is created by myself)

christophKaiser committed:
Added unit test of basin-advanced-brep.ifc
(official example file from buildingsmart.org)

christophKaiser committed:
Added unit test of cube-advanced-brep.ifc
(official example file from buildingsmart.org)

christophKaiser committed:
Updated SupportedIFCrepresentations.md

christophKaiser committed:
Merge branch 'development' into 60AdvancedBrep

christophKaiser committed:
Adjusted unit test of IfcFace: Simplified ifc-structure.
It's intended to test the IfcFace entities (with their IfcFaceOuterBound and IfcFaceBound); thus IfcFacetedBrep and IfcClosedShell aren't necessary for this test.
The images of the visual unit test have not changed (because the geometry hasn't changed).

pjanck commented on Core/src/IfcGeometryConverter/CurveConverter.h:
```diff
@@ -1685,24 +1690,26 @@ namespace OpenInfraPlatform
-					for (auto &orientedEdge : edgeLoop->EdgeList) 
-					{
+					// ToDo: maybe, variable definition of 'loop' has to be moved to correct place
+					std::vector<carve::geom::vector<3>> loop;
```
This todo seems resolved.

christophKaiser replied:
Originaly by change from return-parameter to return value, I wanted to indicate that the declaration and return was just (roughly) put at the begin and end of the function - I haven't looked carefully into details. However, I will rework this function in the refactoring of issue #528 in a few weeks.
For now, the ToDo is removed with commit 314fe78.

pjanck commented on Core/src/IfcGeometryConverter/CurveConverter.h:
```diff
@@ -1748,6 +1755,9 @@ namespace OpenInfraPlatform
 					}
+
+					// ToDo: maybe, return has to be moved to correct place
+					return loop;
```
This todo seems resolved.

christophKaiser replied:
Originaly by change from return-parameter to return value, I wanted to indicate that the declaration and return was just (roughly) put at the begin and end of the function - I haven't looked carefully into details. However, I will rework this function in the refactoring of issue #528 in a few weeks.
For now, the ToDo is removed with commit 314fe78.

pjanck commented on Core/src/IfcGeometryConverter/FaceConverter.h:
```diff
@@ -726,384 +726,279 @@ namespace OpenInfraPlatform {
+					const std::vector<std::vector<std::shared_ptr<carve::input::PolyhedronData>>>& surfacePolyhedronsCollection{
+						itemDataSurface->closed_polyhedrons,
+						itemDataSurface->open_polyhedrons,
+						itemDataSurface->open_or_closed_polyhedrons };
```
I fear that this calls copy-constructor every time the function is called.

Consider making this  `std::vector<std::vector<std::shared_ptr<carve::input::PolyhedronData>>*>`, as in: an array of pointers to ...

christophKaiser replied:
Done in commit c646860.

pjanck commented on Documentation/markdown/SupportedIFCrepresentations.md:
```diff
@@ -72,8 +72,8 @@ Meaning of columns:
-| `IfcFaceBound`                        | :heavy_check_mark:   | `FaceConverter`            | :x:                   | -     |
-| `IfcFaceSurface`                      | :x:                  | -                          | -                     | -     |
+| `IfcFaceBound`                        | :heavy_check_mark:   | `FaceConverter`            | :heavy_check_mark:    | -     |
+| `IfcFaceSurface`                      | :heavy_check_mark:   | -                          | -                     | -     |
```
```suggestion
| `IfcFaceSurface`                      | :heavy_check_mark:   | `FaceConverter`            | -                     | -     |
```

christophKaiser replied:
Done in commit f621469.

pjanck reviewed (approved):
Great work!

Small suggestions, but this can definitely be merged.

jschlenger reviewed (approved):
I believe we can merge after the last few comments of pjanck are resolved

christophKaiser committed:
Fixed variable name.
(Small change to be more understandable - it confused me during the current rework of this function)

christophKaiser committed:
Changed vector to hold pointers to polyhedrons.

christophKaiser committed:
Removed ToDo's in convertIfcEdgeLoop

christophKaiser committed:
Fixed missing Converter-entry in list of supported ifc-entities

pjanck reviewed (approved):
Awesome work!
